polyply/src/solvator.py

```diff
@@ -0,0 +1,138 @@
+import scipy
+import scipy.optimize
+from tqdm import tqdm
+from .processor import Processor
+from .nonbond_engine import POTENTIAL_FUNC
+from .random_walk import fulfill_geometrical_constraints
+
+def norm_lennard_jones_force(dist, sig, eps, force):
+    """
+    Norm of the LJ force between two particles.
+    """
+    return np.abs(24 * eps / dist * ((2 * (sig/dist)**12.0) - (sig/dist)**6)) - force
+
+class Solvator(Processor):
+    """
+    Generate coordinates for 1 residue
+    molecules that are often solvent
+    molecules.
+    """
+    def __init__(self,
+                 nonbond_matrix,
+                 max_force,
+                 mol_idxs,
+                 potential="LJ",
+                ):
+        """
+        Parameters
+        ----------
+        nonbond_matrix: :class:`polyply.src.nonbond_engine.NonBondMatrix`
+            the nonbonded forces and positions of the system
+        max_force: float
+            maximum allowed force on a particle in the one bead per residue
+            description
+        mol_idxs:
+            the indices of the solvent molecules to build
+        potential:
+            the form of the nonbonded potential
+        """
+        self.nonbond_matrix = nonbond_matrix
+        # concatenate all trees into one
+        self.nonbond_matrix.concatenate_trees()
+        # for convenience
+        self.box = self.nonbond_matrix.boxsize
+        self.max_force = max_force
+        self.mol_idxs = np.array(mol_idxs, dtype=int)
+        self.potential = potential
+
+    def clever_grid(self):
+        """
+        Given the largest sigma value of a collection of one bead per residue
+        description of solvents generate a grid that has as miniumum grid spacing
+        the distance at which the LJ force is at most the maximum force-criterion
+        provided. This also works for a mixed system of solvents and sizes, because
+        all cross interactions are geometric averages meaning they cannot exceed
+        the size of the self-interaction.
+        """
+        # find the largest self interaction sigma between two of the solvent beads
+        max_sigma = 0
+        for idx in self.mol_idxs:
+            gndx = self.nonbond_matrix.nodes_to_gndx[(idx, 0)]
+            atype_key = frozenset([self.nonbond_matrix.atypes[gndx]])
+            sigma = self.nonbond_matrix.interaction_matrix[atype_key][0]
+            if sigma > max_sigma:
+                max_sigma = sigma
+        # compute the minimum distance these beads need to have in order
+        # to fullfill the max-force criterion
+        min_dist = scipy.optimize.fsolve(norm_lennard_jones_force,
+                                         x0=0.2,
+                                         args=(max_sigma, 1.0, self.max_force))
+        print(min_dist)
+        # build grid
+        grid = np.mgrid[0:self.box[0]:min_dist,
+                        0:self.box[1]:min_dist,
+                        0:self.box[2]:min_dist].reshape(3, -1).T
```

Comment:
Cubic grid. Did you consider e.g. hexagonal packing instead?

Reply:
Yes but cubic grids are only supported by cKD tree at the moment. I found a numba rewrite of KD tree though but replacing that would screw performance for all non numba users. And I don't understand enough Cython to change the KDtree directly

Reply:
Plus not many tools handle non cubic PBC. It's really less common than you'd think

Reply:
No, I mean that the solvent you build here is placed on cubic gridpoints (due to mgrid). You can also place them on hexagonal gridpoints in a cubic pbc box. How you place solvent molecules has nothing to do with PBC. Okay, apart from some edge effects maybe

Reply:
does that have any advantage?

Reply:
Better packing. Whether that's desirable I don't know.

Reply:
I haven't had any problems with cubic. Changing to hexagonal only introduces PBC artifacts. Previously all packing was one cubic as well. Seems to work well practically. Unless we come up with a specific use case, I wouldn't want to change this here.